src/core/thread/mle.cpp

```diff
@@ -824,30 +824,27 @@ Error Mle::SetDeviceMode(DeviceMode aDeviceMode)
 
     IgnoreError(Store());
 
-    switch (mRole)
+    // We need to re-attach on switching between MTD/FTD modes
+    // and also on switching from rx-on to sleepy (rx-off) mode.
+
+    if (IsAttached() && ((oldMode.IsFullThreadDevice() != mDeviceMode.IsFullThreadDevice()) ||
+                         (oldMode.IsRxOnWhenIdle() && !mDeviceMode.IsRxOnWhenIdle())))
```

Comment:
The updated spec does allow a device to migrate from rx-on to rx-off as long as it initially attached as rx-off. However, this capability is a MAY in the spec. Do we want to support this case?